### PR TITLE
TCP refactoring

### PR DESCRIPTION
**Pull request by pfl**
This PR will move TCP functionality from net_context.c to tcp.c, while leaving TCP function calls in net_context.c that resolve to empty inline functions when TCP is not enabled. As small pieces as possible are worked on one by one, except for the last patch that moves a majority of the TCP functionality to the new file. No functionality has been changed, except for patches 4/14 and 13/14. The issue described this desired change is #5482.

Patch 4/14 reorders the function calls, so that address checks happen before offloading, and any bind calls to the stack happen after offloading.
Patch 13/14 contains a fix where calling connect will set both local and remote addresses also for a UDP.

### CODE REVIEW

File: subsys/net/ip/tcp.c [hunk 1]
```diff
@@ -1344,3 +1412,1160 @@ int net_tcp_parse_opts(struct net_pkt *pkt, int opt_totlen,
 	NET_ERR("Invalid TCP opt: %d len: %d", opt, optlen);
 	return -EINVAL;
 }
+
+int tcp_hdr_len(struct net_pkt *pkt)
```
Comment (lpereira) [inline subsys/net/ip/tcp.c]: This function used to be static; if it's not any longer, please call it `net_tcp_hdr_len()`.

File: subsys/net/ip/tcp.c [hunk 2]
```diff
@@ -1344,3 +1412,1160 @@ int net_tcp_parse_opts(struct net_pkt *pkt, int opt_totlen,
...
+		return -ENOTCONN;
+	}
+
+	return -EOPNOTSUPP;
+}
+
+int net_tcp_listen(struct net_context *context)
+{
+	if (net_context_get_ip_proto(context) == IPPROTO_TCP) {
+		net_tcp_change_state(context->tcp, NET_TCP_LISTEN);
```
Comment (lpereira) [inline subsys/net/ip/tcp.c]: Is `net_tcp_change_state()` exported? If so, can it be used only within `tcp.c`, and marked as static? Same idea for `net_tcp_get_state()`.

**Review by lpereira — COMMENTED** [review]
I wanted to do this for a long time, so thank you for working on this.  It's a big patch and it's been a while since I've worked on the network stack, so I only have a few comments.

File: subsys/net/ip/net_context.c [hunk 3]
```diff
@@ -399,65 +399,74 @@ int net_context_get(sa_family_t family,
 
 #if defined(CONFIG_NET_CONTEXT_CHECK)
 
+	switch(family) {
```
Comment (tbursztyka) [inline subsys/net/ip/net_context.c]: Why changing that to a switch?

File: subsys/net/ip/net_context.c [hunk 4]
```diff
@@ -399,65 +399,74 @@ int net_context_get(sa_family_t family,
...
 #endif
+		break;
+
+	default:
+		NET_ASSERT_INFO(family == AF_INET || family == AF_INET6,
+				"Unknown address family %d", family);
+		return -EAFNOSUPPORT;
+	}
 
+	switch(type) {
```
Comment (tbursztyka) [inline subsys/net/ip/net_context.c]: same as above, why moving towards a switch here?

File: subsys/net/ip/net_context.c [hunk 5]
```diff
@@ -399,65 +399,74 @@ int net_context_get(sa_family_t family,
...
 #if !defined(CONFIG_NET_IPV6)
-	if (family == AF_INET6) {
 		NET_ASSERT_INFO(family != AF_INET6, "IPv6 disabled");
 		return -EPFNOSUPPORT;
-	}
 #endif
+		break;
+
+	default:
+		NET_ASSERT_INFO(family == AF_INET || family == AF_INET6,
```
Comment (tbursztyka) [inline subsys/net/ip/net_context.c]: why retesting this as it is down where family cannot be AF_INET or AF_INET6 ?
Just NET_ERR/NET_WARN here no?

File: subsys/net/ip/net_context.c [hunk 6]
```diff
@@ -399,65 +399,74 @@ int net_context_get(sa_family_t family,
...
+				"Unknown address family %d", family);
+		return -EAFNOSUPPORT;
+	}
 
+	switch(type) {
+	case SOCK_DGRAM:
 #if !defined(CONFIG_NET_UDP)
-	if (type == SOCK_DGRAM) {
 		NET_ASSERT_INFO(type != SOCK_DGRAM,
```
Comment (tbursztyka) [inline subsys/net/ip/net_context.c]: ok so legacy code is using an ASSERT where NET_ERR/NET_WARN could have been sufficient.

File: subsys/net/ip/net_context.c [hunk 7]
```diff
@@ -399,65 +399,74 @@ int net_context_get(sa_family_t family,
...
-		NET_ASSERT_INFO(ip_proto != IPPROTO_UDP, "UDP disabled");
-		return -EPROTONOSUPPORT;
-	}
 #endif
+		break;
 
+	case SOCK_STREAM:
 #if !defined(CONFIG_NET_TCP)
-	if (type == SOCK_STREAM) {
 		NET_ASSERT_INFO(type != SOCK_STREAM,
```
Comment (tbursztyka) [inline subsys/net/ip/net_context.c]: same here

File: subsys/net/ip/net_context.c [hunk 8]
```diff
@@ -399,65 +399,74 @@ int net_context_get(sa_family_t family,
...
 
-	if (family != AF_INET && family != AF_INET6) {
-		NET_ASSERT_INFO(family == AF_INET || family == AF_INET6,
-				"Unknown address family %d", family);
-		return -EAFNOSUPPORT;
-	}
-
-	if (type != SOCK_DGRAM && type != SOCK_STREAM) {
+	default:
 		NET_ASSERT_INFO(type == SOCK_DGRAM || type == SOCK_STREAM,
```
Comment (tbursztyka) [inline subsys/net/ip/net_context.c]: same

File: subsys/net/ip/net_context.c [hunk 9]
```diff
@@ -399,65 +399,74 @@ int net_context_get(sa_family_t family,
...
 				"Unknown context type");
 		return -EPROTOTYPE;
 	}
 
-	if (ip_proto != IPPROTO_UDP && ip_proto != IPPROTO_TCP) {
+	switch (ip_proto) {
```
Comment (tbursztyka) [inline subsys/net/ip/net_context.c]: again, why moving towards switch?

File: subsys/net/ip/net_context.c [hunk 10]
```diff
@@ -2217,17 +2217,20 @@ static void set_appdata_values(struct net_pkt *pkt, enum net_ip_protocol proto)
 	struct net_buf *frag;
 	u16_t offset;
 
+	switch(proto) {
```
Comment (tbursztyka) [inline subsys/net/ip/net_context.c]: again, why moving to a switch here?

**Review by tbursztyka — COMMENTED** [review]
Original code was not using switch on purpose: so it would really get rid of an unsupported family test depending on Kconfig option. This was saving some rom. It's not much locally, but once, twice, ... n times: it saves bits. (note that a  switch on 2 choices isn't even faster than if/else, surely it's nicer to read but that's it).

So either you keep legacy code as is, or you still use #ifdef/#endif relevantly to nuke the unsupported cases from the switch (compiler will then optimize properly).

File: subsys/net/ip/tcp.c (hunk 1 above)
Comment (pfl) [reply subsys/net/ip/tcp.c]: This function became public, as it was moved to tcp.c. With the rest of the stack moving to tcp.c, this function can again be made static.

**Comment by pfl** [comment]
Two bugs discovered and fixed, I'll update this and address the comments.

File: subsys/net/ip/tcp.c (hunk 1 above)
Comment (pfl) [reply subsys/net/ip/tcp.c]: Sorry, no. tcp_hdr_len() is defined in tcp.c, but the usage is actually in net_context.c.

**Review by jukkar — APPROVED** [review]
LGTM

File: tests/net/socket/tcp/testcase.yaml [hunk 11]
```diff
@@ -1,7 +1,4 @@
 tests:
   test:
-    extra_configs:
```
Comment (pfalcon) [inline tests/net/socket/tcp/testcase.yaml]: The idea here was to make sure that the test works with both "normal" setup (the same as the samples are run with) and with CONFIG_NET_TEST=y. When something is made to work with only non-default setup, there's always a suspicion that some black voodoo is needed to make it work *at all*.

**Pull request merged**